### PR TITLE
Let non-users unsubscribe from alerts and subscriptions

### LINKED ISSUE #29955: Non-Metabase users are unable to unsubscribe from alerts they have been subscribed to
[Product doc](https://www.notion.so/metabase/Let-non-users-unsubscribe-from-alerts-and-subscriptions-efc38b07283944f3b084f84146fed7ca)

**Is your feature request related to a problem? Please describe.**
Subscription recipients who are not Metabase users are unable to unsubscribe.

**Describe the solution you'd like**
- Allow subscription recipients who are not Metabase users to unsubscribe

**Describe alternatives you've considered**
- Add a subscription confirmation email prior to subscribing non-users to alerts

**How important is this feature to you?**
- Requested by customer
- This is potentially dangerous issue as a typo could cause the sender's email host (`.metabaseapp.com` or customer's host) to be marked as spam or damage the sending server reputation which is very difficult to fix.

**Additional context**
image

```[tasklist]
### Tasks
- [x] (BE) API Route and Backend Functionality (https://github.com/metabase/metabase/pull/30597)
- [x] (BE) Email Notifcation URL for Non-Users (https://github.com/metabase/metabase/pull/30892)
- [x] (BE) Return pulse name with unsubscribe request (https://github.com/metabase/metabase/pull/31672)
- [x] (FE) Create Unsubscribe Page and API Requests (https://github.com/metabase/metabase/pull/32379)
```

### PR DESCRIPTION
**Pull request by JesseSDevaney**
**Problem:** Non-users who are subscribed to dashboard subscriptions have no direct way of unsubscribing themselves.

**Context:** #29955 

---
**Desired Behavior**/**Acceptance Criteria**

_Successful Unsubscribe_

https://github.com/metabase/metabase/assets/22608765/897a2638-a730-4a20-a61e-08e83a71518e

- [ ] Upon receiving a subscription, a non-user should have an `Unsubscribe` link at the bottom of their email (with valid parameters)
- [ ] Clicking on the `Unsubscribe` link should navigate the non-user to the Unsubscribe page (`/unsubscribe?[hash]&[email]&[pulse-id]`) and should automatically unsubscribe them from the subscription
- [ ] Upon unsubscribing, the user should be given the one-time option to undo the unsubscribe in case they misclicked the unsubscribe button. Clicking the `Undo` button should resubscribe them to the subscription

_404 Page: Missing Required Parameters_ (`hash`, `email`, or `pulse-id`)

404-page

- [ ] If a user navigates to the unsubscribe page without supplying the required parameters, show them a 404 page

_Error Message: Server Error_ (invalid parameters, random server error)

error-message

- [ ] If a user navigates with invalid paramters (invalid hash, non-subscribed email, non-valid pulse-id), show them an error message
- [ ] If the server responds with a random error, show them an error message

_Solution Validation_
- [ ] Tests added to validate and enforce the desired behavior
  - [ ] Successful unsubscribe of non-user
  - [ ] Successful undo-unsubscribe of non-user
  - [ ] 404 page displayed for not supplying the required parameters
  - [ ] Error message for server errors
- [ ] All other tests continue to pass

---
**Development Setup**
1. Spin up a [mail server](https://hub.docker.com/r/maildev/maildev) to receive subscription emails
2. Setup [Metabase emailing](https://www.metabase.com/docs/latest/configuring-metabase/email) for your local environment
3. Create a dashboard
4. Add a scalar question
5. Setup subscriptions
    - Navigate to subscriptions > add non-user email (e.g. "non-user@example.com") > click `Done`
    - _Note: You must click `Done` before clicking `Send email now` or else the pulse will not be setup properly and the unsubscribe URL will be broken (not have a valid pulse-id)_
6. Navigate back to subscriptions and click `Send email now`
7. Navigate to email that was sent and see the `Unsubscribe` link at the bottom of the email

### CODE REVIEW

**Review by vbenedetti — APPROVED** [review]
Looks good to me and the solution is complete with error cases.

File: frontend/src/metabase/containers/Unsubscribe.tsx [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+  LayoutIllustration,
+  LayoutRoot,
+  CheckmarkIcon,
+} from "metabase/containers/Unsubscribe.styled";
+
+const ERRORS = {
+  MISSING_REQUIRED_PARAMETERS: "missing required parameters",
+};
+
+const SUBSCRIPTION = {
```
Comment (EmmadUsmani) [inline frontend/src/metabase/containers/Unsubscribe.tsx:29]: Nit: There is a slightly more typesafe way of writing this. Right now the values are treated as string types, so I can call `setSubscriptionChange("any string value")` without a type error.

If we use `as const` and create a `Subscription` type from the object, then we can use that type to ensure the state value can only be `unsubscribe` or `resubscribe`. Same goes for the `ERRORS` object as well.

```typescript
const SUBSCRIPTION = {
  UNSUBSCRIBE: "unsubscribe",
  RESUBSCRIBE: "resubscribe",
} as const;
type Subscription = typeof SUBSCRIPTION[keyof typeof SUBSCRIPTION];
...
  const [subscriptionChange, setSubscriptionChange] = useState<Subscription>(
    SUBSCRIPTION.UNSUBSCRIBE,
  );
  ```

**Review by EmmadUsmani — APPROVED** [review]
Overall looks good! Left a few non-blocking commentss

File: frontend/src/metabase/home/components/HomeLayout/HomeLayout.tsx [hunk 2]
```diff
@@ -4,7 +4,9 @@ import { getUserIsAdmin } from "metabase/selectors/user";
 import MetabotWidget from "metabase/metabot/components/MetabotWidget";
 import Tooltip from "metabase/core/components/Tooltip/Tooltip";
 import { HomeGreeting } from "../HomeGreeting";
+// ! NEED TO USE THIS FOR ILLUSTRATION AS WELL
```
[the remark this answers is not in the document]
Comment (JesseSDevaney) [reply frontend/src/metabase/home/components/HomeLayout/HomeLayout.tsx:7]: We very much do not need to keep this comment. Thank you for catching this :sweat_smile:

File: frontend/src/metabase/dashboard/components/DashboardActions.jsx [hunk 3]
```diff
@@ -55,7 +55,7 @@ export const getDashboardActions = (
             icon="subscription"
             disabled={!canManageSubscriptions}
             onClick={onSharingClick}
-            data-metabase-event="Dashboard;Subscriptions"
```
[the remark this answers is not in the document]
Comment (JesseSDevaney) [reply frontend/src/metabase/dashboard/components/DashboardActions.jsx:58]: When I added an aria-label for this button, I accidentally deleted this line. I am adding a commit to undo this deletion.

File: frontend/src/metabase/containers/Unsubscribe.tsx (hunk 1 above)
Comment (JesseSDevaney) [reply frontend/src/metabase/containers/Unsubscribe.tsx:29]: Thank you very much for this - very neat! :heart:

**Pull request merged**